poc/flp_generic.py

```diff
@@ -640,6 +641,140 @@ def test_vec_set_type_param(self, test_vec):
+    - Client will encode `count = offset + \sum_i x_i` in `num_bits_for_count`
+      bits.
+    - We can naturally bound `count` as the following:
+      `0 <= count <= 2**num_bits_for_count - 1`, and therefore:
+      `-offset <= \sum_i x_i <= max_num_bits`.
+    - Since we also verify each `x_i` is a bit, we can lower bound the summation
+      by 0. Therefore, we will be able to verify
+      `0 <= \sum_i x_i <= max_num_bits`.
+    """
+    # Operational parameters
+    length = None  # Set by `MultiHotHistogram.with_params()`
+    max_num_bits = None  # Set by `MultiHotHistogram.with_params()`
+    chunk_length = None  # Set by `MultiHotHistogram.with_params()`
+
+    # Associated types
+    Measurement = Unsigned
+    AggResult = Vec[Unsigned]
+    Field = field.Field128
+
+    # Associated parameters
+    GADGETS = None  # Set by `MultiHotHistogram.with_params()`
+    GADGET_CALLS = None  # Set by `MultiHotHistogram.with_params()`
+    MEAS_LEN = None  # Set by `MultiHotHistogram.with_params()`
+    JOINT_RAND_LEN = 2
+    OUTPUT_LEN = None  # Set by `MultiHotHistogram.with_params()`
+
+    def __init__(self, length, max_num_bits, chunk_length):
+        """
+        Instantiate an instance of the `MultiHotHistogram` circuit with the
+        given length, max_num_bits, and chunk_length.
+        """
+        if length <= 0:
+            raise ValueError('invalid length')
+        if max_num_bits <= 0 or max_num_bits >= length:
+            raise ValueError('invalid max_num_bits')
+        if chunk_length <= 0:
+            raise ValueError('invalid chunk_length')
+
+        # Compute the number of bits to represent `max_num_bits`.
+        self.num_bits_for_count = int(math.log2(max_num_bits)) + 1
+        self.offset = self.Field(
+            (1 << self.num_bits_for_count) - 1 - max_num_bits
+        )
+        # Sanity check `offset + length` doesn't overflow field size,
+        # because in validity circuit, we will compute `offset + \sum_i x_i`.
+        if self.Field.MODULUS - self.offset.as_unsigned() <= length:
+            raise ValueError('length and max_num_bits are too large '
+                             'for the current field size')
+
+        self.length = length
+        self.max_num_bits = max_num_bits
+        self.chunk_length = chunk_length
+        self.GADGETS = [ParallelSum(Mul(), chunk_length)]
+        # The number of bit entries are `length + num_bits_for_count`,
+        # so the number of gadget calls is equal to
+        # `ceil((length + num_bits_for_count) / chunk_length)`.
+        self.GADGET_CALLS = [
+            (length + self.num_bits_for_count + chunk_length - 1) //
+            chunk_length
+        ]
+        self.MEAS_LEN = self.length + self.num_bits_for_count
+        self.OUTPUT_LEN = self.length
+
+    def eval(self, meas, joint_rand, num_shares):
+        self.check_valid_eval(meas, joint_rand)
+
+        # Check that each bucket is one or zero.
```

Comment:
Is this sufficient for the DP use case? Suppose our algorithm for adding noise is something like this:
```python3
for bucket_index in len(meas):
   meas[bucket_index] += sample_independent_noise()
```
Depending on what `sample_independent_noise()` does, `meas[bucket_index]` might be something other than `0` or `1`.

Reply:
Great, thanks for clarifying.